src/esp/sensor/FisheyeSensor.cpp

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "FisheyeSensor.h"
+
+#include <Corrade/Utility/Assert.h>
+#include <Corrade/Utility/FormatStl.h>
+#include "esp/gfx/DoubleSphereCameraShader.h"
+
+namespace Mn = Magnum;
+namespace Cr = Corrade;
+
+namespace esp {
+namespace sensor {
+
+FisheyeSensorSpec::FisheyeSensorSpec() : VisualSensorSpec() {
+  uuid = "fisheye";
+  sensorSubType = SensorSubType::Fisheye;
+}
+
+void FisheyeSensorSpec::sanityCheck() {
+  VisualSensorSpec::sanityCheck();
+  CORRADE_ASSERT(
+      sensorSubType == SensorSubType::Fisheye,
+      "FisheyeSensorSpec::sanityCheck(): sensorSpec is not Fisheye", );
+
+  CORRADE_ASSERT(focalLength[0] > 0 && focalLength[1] > 0,
+                 "FisheyeSensorSpec::sanityCheck(): focal length,"
+                     << focalLength << "is illegal.", );
+}
+
+void FisheyeSensorDoubleSphereSpec::sanityCheck() {
+  FisheyeSensorSpec::sanityCheck();
+  CORRADE_ASSERT(fisheyeModelType == FisheyeSensorModelType::DoubleSphere,
+                 "FisheyeSensorDoubleSphereSpec::sanityCheck(): fisheye model "
+                 "type is wrong. It should be DoubleSphere.", );
+  CORRADE_ASSERT(alpha >= 0 && alpha <= 1.0,
+                 "FisheyeSensorDoubleSphereSpec::sanityCheck(): alpha"
+                     << alpha << "is illegal.", );
+}
+
+template <typename T>
+void specSanityCheck(FisheyeSensorSpec* spec) {
+  auto actualSpec = dynamic_cast<T*>(spec);
+  CORRADE_ASSERT(actualSpec,
+                 "FisheyeSensor::FisheyeSensor(): the spec cannot be converted "
+                 "to any known fiesheye sensor spec.", );
+  actualSpec->sanityCheck();
+}
+
+FisheyeSensor::FisheyeSensor(scene::SceneNode& cameraNode,
+                             const FisheyeSensorSpec::ptr& spec)
+    : VisualSensor(cameraNode, spec) {
+  switch (fisheyeSensorSpec_->fisheyeModelType) {
+    case FisheyeSensorModelType::DoubleSphere: {
+      specSanityCheck<FisheyeSensorDoubleSphereSpec>(fisheyeSensorSpec_.get());
+    } break;
+  };
+
+  // initialize a cubemap
+  auto& res = fisheyeSensorSpec_->resolution;
+  int size = res[0] < res[1] ? res[0] : res[1];
```

Comment:
We talked about this in a recent meeting and I'm thinking about this more. What if the fisheye FoV is very small (i.e., large focal length)? It seems like the cubemap needs to be higher resolution when the fisheye FoV is smaller.

I would love to see some sample images at different focal lengths.

Reply:
> It seems like the cubemap needs to be higher resolution when the fisheye FoV is smaller.

May I understand why.

> I would love to see some sample images at different focal lengths.

I will try to create them.

Reply:
>May I understand why.

If the FOV were very small, the fisheye sensor would be reading all the pixels for its sensor from a very small portion of the cubemap. If there aren't enough texels in the cubemap, the fisheye render will look blurry.

Reply:
@Skylion007 : You are using the fisheye sensor for training. Do you have any opinions here?

Reply:
>If the FoV were very small, the fisheye sensor would be reading all the pixels for its sensor from a very small portion of the cubemap. If there aren't enough texels in the cubemap, the fisheye render will look blurry.

Now I understand your concern. Thanks. Here are my thoughts.

First, fisheye sensor is designed to handle the large FOV. If FOV is small, we use pinhole camera model directly. 

Second, let us look into this problem:
The process to render the observation of a fisheye sensor is similar to the ray tracing. For each pixel (u,v) in the image space, we try to unproject it to a ray (2D -> 3D, all points on this ray will be projected to this (u, v)). We sample the cubemap using this ray, a unit vector, and thus find the "light source" pixel on the cubemap.

Blur is more likely to happen to the pixels that are closer to the principal axis. Because a neighboring pixel pair closer to the principal axis will be unprojected to 2 rays that have smaller included angle. That means it has higher chance to hit the same pixel on the cubemap.

**One solution I was thinking** (I did not implement it in this PR due to complexity.):
We know (or compute) the `FOV`, which is an angle. And we can compute the smallest included `angel` between rays, which are projected from the pixels `(h/2 + 1, w/2)` and `(h/2, w/2)` (assuming the principle axis goes through the image center, `(h/2, w/2)`) Then the resolution of the cubemap can be computed as: <img src="https://render.githubusercontent.com/render/math?math=\lceil{\frac{FOV}{angle}}\rceil">. It guarantees 1 pixel offset on the image leads to at least 1 pixel offset on the cubemap.

I was asking why, because I never worried about the resolution of cubemap would not be high enough. Actually I was concerned the opposite case, that is, the current resolution is more than enough in most cases, which slows down the rendering.

I can try what I proposed above. Since in the "double sphere" camera model, the `FOV` is not given, I need some extra time to derive it. It will further delay this diff. So I prefer to try it in another PR as future optimization.

**One practical solution to address your concern:**
- I will add a `cubemapSize` in the `FisheyeSensorSpec`, and a flag `CustomizedCubeMapSize` in the `FisheyeSensor` class. When the flag is set, the program will adopt the user input.
- If the flag is not set, by default, the cubemap will use `min(w, h)` as I implemented here.